src/ucs/sys/event_set.c

```diff
@@ -134,31 +135,44 @@ ucs_status_t ucs_event_set_del(ucs_sys_event_set_t *event_set, int event_fd)
     return UCS_OK;
 }
 
-ucs_status_t ucs_event_set_wait(ucs_sys_event_set_t *event_set, int timeout_ms,
+ucs_status_t ucs_event_set_wait(ucs_sys_event_set_t *event_set,
+                                unsigned max_events, int timeout_ms,
                                 ucs_event_set_handler_t event_set_handler,
-                                void *arg)
+                                void *arg, unsigned *read_events)
 {
-    int nready;
     struct epoll_event ep_events[UCS_EVENT_EPOLL_MAX_EVENTS];
-    int i;
+    unsigned i, max_wait_events;
+    int nready, events;
 
-    if (event_set_handler == NULL) {
-        return UCS_ERR_INVALID_PARAM;
-    }
+    ucs_assert(event_set_handler != NULL);
+    ucs_assert(read_events != NULL);
 
-    nready = epoll_wait(event_set->epfd, ep_events, UCS_EVENT_EPOLL_MAX_EVENTS,
+    max_wait_events = ucs_min(max_events, UCS_EVENT_EPOLL_MAX_EVENTS);
+    nready = epoll_wait(event_set->epfd, ep_events, max_wait_events,
                         timeout_ms);
-    if ((nready < 0) && (errno != EINTR)) {
+    if (nready < 0) {
+        *read_events = 0;
+
+        if (errno == EINTR) {
+            return UCS_INPROGRESS;
+        }
         ucs_error("epoll_wait() failed: %m");
         return UCS_ERR_IO_ERROR;
     }
+
     ucs_trace_data("epoll_wait(epfd=%d, timeout=%d) returned %d",
                    event_set->epfd, timeout_ms, nready);
 
-    for (i=0; i < nready; i++) {
-        int events = ucs_event_set_map_to_events(ep_events[i].events);
-        event_set_handler(ep_events[i].data.fd, events, arg);
+    for (i = 0; i < nready; i++) {
```

Comment:
there is signed/unsigned compare here (some compilers may not like it)

Reply:
Changed `i` from `unsigned` to `signed`.